### ISSUE TITLE
Fix Iframe loading

### ISSUE THREAD

**Issue by gterzian** [servo/servo#31973]
As part of debugging at new failing test in https://github.com/servo/servo/pull/31505(), I noted some(known?) problems with how iframes are loaded.

The test in question is `/html/browsers/browsing-the-web/navigating-across-documents/initial-empty-document/load-pageshow-events-iframe-contentWindow.html`, which asserts that **no load events are fired** for an iframe which has no navigable("load" & "pageshow" events do not fire on contentWindow of iframe that stays on the initial empty document). 

We fail all tests, meaning we fire load events, for all subtests, with [the exception of](https://github.com/servo/servo/blob/da3991c8f37251e04ba67dd7e71c3129327dddae/tests/wpt/meta/html/browsers/browsing-the-web/navigating-across-documents/initial-empty-document/load-pageshow-events-iframe-contentWindow.html.ini#L4) "load & pageshow events do not fire on contentWindow of <iframe> element created with src=''". So, for some reason we do not fire the load event, at least not within 50ms, in that case(it's not clear to me why). The test fails now also with https://github.com/servo/servo/pull/31505, hence my investigation, but I think I may want to let it fail, and instead fix all tests in one go.

Quick recap on how this relates to the spec:

1. Entry point for processing are [iframe HTML element insertion steps](https://html.spec.whatwg.org/multipage/#the-iframe-element:html-element-insertion-steps), which call into [process the iframe attributes](https://html.spec.whatwg.org/multipage/#process-the-iframe-attributes).
2. We then hit the "2 otherwise"(meaning it has a url), which then either:
3. Immediately returns if there is no src, which means not firing any load event. 
4. Immediately run the load steps for an "about:blank" iframe, which means **firing a load event on the element, not the content window**
5. Navigate the iframe otherwise, which means eventually firing the load event on the content window as per [completely-finish-loading](https://html.spec.whatwg.org/multipage/document-lifecycle.html#completely-finish-loading)

So, in other words, there are three:
- in the case of no src, no event should fire at all, and no navigation should happen. 
- in the case of "about:blank", a load event should synchronously fire on the element, but no navigation should happen. 
- Otherwise, we should navigate the iframe, and then only fire the load event when the load completes(so that's asynchronous). 

-------

What we are currently doing, and why it is wrong:

[bind_to_tree](https://github.com/servo/servo/blob/da3991c8f37251e04ba67dd7e71c3129327dddae/components/script/dom/htmliframeelement.rs#L710) is the entry point, where we do two things via a delayed task:

1. `create_nested_browsing_context`
2. `process_the_iframe_attributes(ProcessingMode::FirstTime)`. 

Both end-up calling into `start_new_pipeline`(with the exception of the no src case, which [returns](https://github.com/servo/servo/blob/da3991c8f37251e04ba67dd7e71c3129327dddae/components/script/dom/htmliframeelement.rs#L292)), where 1 uses `PipelineType::InitialAboutBlank`, and 2 `PipelineType::Navigation`. 

`PipelineType::InitialAboutBlank` results in a kind of synchronous loading of the iframe doc via [`ScriptThread::process_attach_layout`](https://github.com/servo/servo/blob/da3991c8f37251e04ba67dd7e71c3129327dddae/components/script/dom/htmliframeelement.rs#L227), which eventually calls into [document.maybe_queue_document_completion()](https://github.com/servo/servo/blob/da3991c8f37251e04ba67dd7e71c3129327dddae/components/script/script_thread.rs#L1698), which then sends a msg to the constellation by way of (`document.notify_constellation_load()`)[https://github.com/servo/servo/blob/da3991c8f37251e04ba67dd7e71c3129327dddae/components/script/dom/document.rs#L2427]. It does not end there: the constellation will then respond with a [`DispatchIFrameLoadEvent`](https://github.com/servo/servo/blob/da3991c8f37251e04ba67dd7e71c3129327dddae/components/constellation/constellation.rs#L3086) message, which will then result in [calling the iframe load event steps](https://github.com/servo/servo/blob/da3991c8f37251e04ba67dd7e71c3129327dddae/components/script/script_thread.rs#L3050).

`PipelineType::Navigation` starts with a [`ScriptMsg::ScriptLoadedURLInIFrame`](https://github.com/servo/servo/blob/da3991c8f37251e04ba67dd7e71c3129327dddae/components/script/dom/htmliframeelement.rs#L239C27-L239C61) sent to the constellation, resulting in what I assume is a "normal" navigation workflow. 

So somehow we need to restructure this, where:

1. When binding to the tree, start the `PipelineType::InitialAboutBlank`(which is not really a good name, just `Initial` would be better) workflow to do it's thing, 
2. but then when it's back in script it should only run the `iframe_load_event_steps` if there is an src. 
3. Then it is only at that point that should we run the `PipelineType::Navigation`, and only if the src is not about:blank.

**Comment by mrobinson** [comment]
@gterzian Very nice exploration! Do you know if this could be causing the thing where we sometimes see two load events when loading `about: blank`.

**Comment by gterzian** [comment]
> Very nice exploration! Do you know if this could be causing the thing where we sometimes see two load events when loading about: blank.

Yes I think so, although there are so many details that it needs some further investigation, but I think if we load `about: blank` we can get two load events: one from the first "synchronous" creation of the BC via `ScriptThread::process_attach_layout`, and one from the subsequent navigation of that BC(since the Iframe "bind to tree" steps start the two workflows).

**Comment by gterzian** [comment]
I've had another chance to debug this as part of https://github.com/servo/servo/pull/32574, and it appears to me that the main issue really is that for each iframe, we always:

1. First create an about:blank document, and load is synchronously. (see [`create_nested_browsing_context()`](https://github.com/servo/servo/blob/bc1c71bd4df9424e4cceed9dc8cc38efc591095b/components/script/dom/htmliframeelement.rs#L726)
2. Then create another document(with a different pipeline id) and navigate it(see [`process_the_iframe_attributes(ProcessingMode::FirstTime)`](https://github.com/servo/servo/blob/bc1c71bd4df9424e4cceed9dc8cc38efc591095b/components/script/dom/htmliframeelement.rs#L727). 

This actually somewhat matches the spec, which:

1. First creates an about:blank child navigable (https://html.spec.whatwg.org/multipage/document-sequences.html#create-a-new-child-navigable)
2. And their, if the src is actually about blank, does not much after, and otherwise navigates it. 

So I think our problem is that:

1. We create two different pipelines; instead we should use the same.
2. That we always create two different pipelines, even if the iframe's src is "about:blank". 

So I think we should simply indeed create an "about:blank" document as the default, and then process the attributes, and if the src(doc) is "about:blank", then simply do the load even steps, and if it is something else, navigate that document(as opposed to create a second one). 

cc @jdm (I saw some comments in the current code originated from you so I assume you were initially involved).

**Comment by jdm** [comment]
Agreed. Those steps were always the intention behind the code.

**Comment by gterzian** [comment]
The test related to https://github.com/servo/servo/issues/32004 is failing intermittently, so perhaps it depends on this after all. 

https://github.com/servo/servo/pull/32574#issuecomment-2244959149

> So the problem is essentially #31973
> 
> The [test](/html/webappapis/update-rendering/child-document-raf-order.html) creates two iframes, but we actually create four documents(2 for each iframe: one "about:blank" and another), and then we remove the "about:blank" ones when the "real" ones finish navigating. I think the timeout is because the raf end-up being lost because the "about:blank" are not returned anymore as part of `iter_iframes`(while there is a time where even though they are not part of `iter_frames`, they are still inside `documents`) and then they the rafs callbacks are not called.

(the above is not comfirned)

**Comment by jdm** [comment]
>1. We create two different pipelines; instead we should use the same.
>2. That we always create two different pipelines, even if the iframe's src is "about:blank".

I created a dead-simple patch to see what happens if we reuse a pipeline id when navigating an about:blank document inside of an iframe, and it definitely looks like Servo's navigation model is designed around each document having a unique pipeline id.

Going back to your points from https://github.com/servo/servo/issues/31973#issue-2220072851:

>So, in other words, there are three:
>
> 1. in the case of no src, no event should fire at all, and no navigation should happen.
> 2. in the case of "about:blank", a load event should synchronously fire on the element, but no navigation should happen.
> 3. Otherwise, we should navigate the iframe, and then only fire the load event when the load completes(so that's asynchronous).

Iframe load events have a long history of the spec not matching what browsers actually do, and what browsers do not matching each other: https://github.com/whatwg/html/issues/4965 and https://hsivonen.fi/about-blank/ lay out the history here.

More thoughts later when I've had time to read this issue more closely.

**Comment by jdm** [comment]
Ok, I've read more (eg. https://github.com/whatwg/html/pull/6869#issue-946630091) and I agree that we should ensure that our implementation matches what is written in the spec and what is present in the tests. That appear to be the model that browsers are either already using or converging upon (see https://github.com/whatwg/html/issues/3267#issuecomment-1232545141).

**Comment by gterzian** [comment]
I'll look into this as well on my side(besides https://github.com/servo/servo/pull/33087), in the least it will make reviewing something easier. We can compare drafts later @jdm

**Comment by sebsebmc** [comment]
I started trying to investigate secure-contexts tests as a part of figuring out #36346 and I see that the current tests are timing out when we set srcdoc on sandboxed iframes and I think it may be related to these issues.  I was able to find that somehow these iframes are having their loading end up in `scheme_fetch` where `about:srcdoc` is not handled.

**Comment by jdm** [comment]
Oh, that's interesting. about:srcdoc should never make it to the network thread.

**Comment by jdm** [comment]
I intend to return to this work soon, and my plan is to attempt to pull apart the changes in #33087 so they can merge incrementally. This will likely require accepting regressions in test results along the way, since each change I made in the original PR uncovered new incomplete/incorrect features.

**Comment by gterzian** [comment]
@jdm myself and @Taym95 are thinking of looking into this one; is that ok with you or are you currently working on it, or planning to?

**Comment by jdm** [comment]
I'm ok with that but we should probably discuss a strategy!

**Comment by gterzian** [comment]
Finally starting to get back into this today.

**Comment by mrobinson** [comment]
@Loirooriol and I have recently done a bit of work on the `sandbox` attribute for `<iframe>`s. If `allow-same-origin` isn't set in the `sandbox` attribute it is important to note that the iframe content should still be loaded in a new `ScriptThread.

**Comment by gterzian** [comment]
> loaded in a new `ScriptThread

Yes the part of how we create or re-use script-threads(event-loops at the constellation), which is essentially at [pipeline creation time](https://github.com/servo/servo/blob/16226aea46a9d253da0d02707ceab15f5832153c/components/constellation/constellation.rs#L961) at the level of the constellation, I haven't been able to dive back in yet. 

My current understanding is that we:

1. First create an about:blank iframe in whatever script-thread the surrounding document is running, and then
2. We navigate that newly created iframe, which creates another iframe. 

And I believe that at 2, we potentially switch event-loops(by that I mean that the second iframe is created to run in another event-loop than the first) and then destroy the previously created iframe.

**Comment by gterzian** [comment]
Adding "closing" to this issue, because the loading and closing problems are closely(no pun intended) related.

**Comment by gterzian** [comment]
Following https://github.com/servo/servo/pull/40100 I can think of a couple of good next steps:

- Look into running the iframe load event steps synchronously, as in not via the handling of [`DispatchIFrameLoadEvent`](https://github.com/servo/servo/blob/9fe53e5369646c485a2b33b54b1f6f0583b17aab/components/constellation/constellation.rs#L3142)(you probably still need to send the `ScriptToConstellationMessage::LoadComplete` message).
- Look into implementing https://html.spec.whatwg.org/multipage/dom.html#is-initial-about:blank (this ties into https://github.com/servo/servo/issues/31982 as well)
- Implement the solution mentioned at https://github.com/servo/servo/issues/28298#issuecomment-3454722828 (when unloading iframes)
